What is Holochain Review/edits Take 2 (#157)
holochain/docs-pages: src/what-is-holochain.md

## How does it compare to...

### Server-side frameworks?

Holochain is similar to frameworks like Django, ASP.NET, Laravel, Express, and Ruby on Rails---it gives you the tools to write a full-featured 'back end' for your application.

Key differences:

* It's opinionated about data storage, providing its own persistence layer.
* Data validation is a required application component; it's implemented in callback functions that execute when data is about to be written.
* Identity is managed via [public key cryptography](https://en.wikipedia.org/wiki/Public-key_cryptography), similarly to SSH. Users create keys on their own machines.
* The back end and the front end both run on the devices of its users, not a server.

### Relational / <abbr title="structured query language">SQL</abbr> databases?

Each Holochain application has its own separate data store, a <abbr title="distributed hash table">DHT</abbr>. Records in the DHT have a defined and enforced schema, similar to a table schema. Data in the DHT is stored differently from a SQL database.

Key differences:

* Data is held by the users of the app on their own devices.

> * Data propagation is eventually consistent. Not all users will see the exact same data at the same time.

Reviewer: ```suggestion
* Data propagation is [eventually consistent](https://en.wikipedia.org/wiki/Eventual_consistency). Not all users will see the exact same data at the same time.
```

> * All public data is visible to anyone with access to the application. However, users can store private data on their own devices, which can't be read by users of other applications.

Reviewer: The "can't be read by users of other applications" was applicable to "all public data" rather than private data. Pulled the line out; don't think it's necessary anymore, because DB designers always assume that their database can't be read by applications :D

```suggestion
* All public data is visible to anyone with access to the application. However, users can also store some private data on their own devices.
```

* Data is not stored in tables, but in individual entries that are retrieved by their unique ID.
* Relations between entries are created with links, making the DHT more like a graph database than a relational database.
* Currently, you can't perform column-based or relational queries like you can with SQL, although limited querying functionality can be built with entries and links. We intend to build more indexing and querying features in the future.

> * Each user authors their own data and stores it in a journal on their own device before publishing it to the DHT. This journal can also hold private entries that are not published to the DHT.

Reviewer: ```suggestion
* Each user writes data to a journal on their own device before publishing it to the DHT.
```

### NoSQL databases?

> Holochain's public graph database is similar to document stores such as MongoDB and Elastic, and key/value stores such as Redis. It's closest most resembles graph databases such as Neo4j, which makes it a good match for [<abbr title="Resource Description Framework">RDF</abbr>](https://www.w3.org/RDF/) data and [GraphQL](https://graphql.org/)- or [SPARQL](https://en.wikipedia.org/wiki/SPARQL)-based data access layers.

Reviewer: ```suggestion
Holochain's public graph database is similar to document stores such as MongoDB and Elastic, and key/value stores such as Redis. It most closely resembles graph databases such as Neo4j, which makes it a good match for [<abbr title="Resource Description Framework">RDF</abbr>](https://www.w3.org/RDF/) data and [GraphQL](https://graphql.org/)- or [SPARQL](https://en.wikipedia.org/wiki/SPARQL)-based data access layers.
```

Key differences:

* Data is held by the users of the app on their own devices.
* All of an application's public data is visible to all users of that application, although users can store private data on their own devices.
* Field-based queries are not yet available, though they can potentially be built using existing data primitives.

### Blockchain and other distributed ledger technologies (<abbr>DLT</abbr>s)?

Holochain could be considered a DLT, since it has some technological and conceptual similarities with blockchain, <abbr title="directed acyclic graph">DAG</abbr> chains, and other DLTs. However, it is founded on radically different assumptions.

Key differences:

* Holochain is not a platform or network, but a development framework (similar to Tendermint).

> * All code execution, data creation, and data storage happens via an individual user running their own node, rather than a "global computer."

Reviewer: ```suggestion
* All code execution, data creation, and data storage happens via an individual user running their own node, rather than a "global computer".
```

;)

* Public data lives in the application's <abbr title="distributed hash table">DHT</abbr>, a semi-structured graph database, rather than a global ledger.
* Holochain has built-in tools to anticipate and support hard fork migrations.
* Data storage and validation responsibility are distributed among nodes via a selection algorithm that creates overlapping 'neighborhoods'. This reduces storage and compute requirements, making Holochain comparable in performance to sharded blockchains or DAG chains.
* Nodes can contact each other directly without writing data to a public ledger or sending it through a gossip network.
* Keys are unique to a node and can't be used on another node. Key management is supported by a built-in distributed <abbr title="public key infrastructure">PKI</abbr> dApp.
* Holochain doesn't natively support anonymity, but is instead optimized for identity continuity.

### Serverless/lambda?

Similarly to serverless offerings, you can use Holochain to deploy an application without provisioning or maintaining server infrastructure. Back end code consists of stateless functions that can call other functions and receive data from external services.

Key differences:

> * Holochain does not provide a serverless abstraction over a cloud infrastructure---in fact, there are no servers at all. Instead, each user hosts both back end and front end code on their own device.

Reviewer: ```suggestion
* Holochain does not create a serverless abstraction on top of cloud servers---there are no servers at all. Instead, each user hosts both back end and front end code on their own device.
```

* Public data is stored in a graph database engine provided by the Holochain runtime.
* Functions are executed only when an individual user says so, and only its local state is affected.  This state may then be optionally published to the public graph database so it can be accessed by others.
* Back end code is typically deployed to interested users as a package of functions. However, a function can call another function in another package if both packages are installed on one user's device.

> ### BitTorrent, Dat, Git, IPFS, Secure Scuttlebutt, and other peer-to-peer (<abbr>P2P</abbr>) or distributed web (<abbr title="distributed web">DWeb</abbr>) technologies?

Reviewer: ```suggestion
### BitTorrent, Dat, Git, GUN, IPFS, Secure Scuttlebutt, and other peer-to-peer (<abbr>P2P</abbr>) or distributed web (<abbr title="distributed web">DWeb</abbr>) technologies?
```

> Holochain is part of a new breed of <abbr>P2P</abbr> <abbr>DWeb</abbr> technologies. It shares the 'agent-centric', offline-friendly, and cryptographically secure approaches of Dat, Git, IPFS, and Secure Scuttlebutt. As a full application development stack with built-in assumptions about user identity, data propagation, and data retrieval, it's most similar to Secure Scuttlebutt.

Reviewer: ```suggestion
Holochain is part of a new breed of <abbr>P2P</abbr> <abbr>DWeb</abbr> technologies. It builds an offline-friendly, 'agent-centric' peer-to-peer network on top of cryptographic primitives, similarly to Dat, Git, IPFS, and Secure Scuttlebutt. As a full application development stack with built-in assumptions about user identity, data propagation, and data retrieval, it's most similar to Secure Scuttlebutt.
```

Key differences:

* Holochain is meant for application development, whereas BitTorrent, Dat, Git, and IPFS are simply data persistence and propagation layers on which applications can be built.

> * Holochain limits which tools may be used for authoring core validation logic, whereas Secure Scuttlebutt allows developers to use their favorite languages and frameworks.

Reviewer: ```suggestion
* Holochain is opinionated about back-end languages and frameworks, whereas Secure Scuttlebutt allows developers to choose their own.
```

'opinionated' has a specific meaning for devs who are shopping around for toolkits, frameworks, and libraries; I'd like to keep this word in here.
